autorest/azure/async_test.go

```diff
@@ -0,0 +1,516 @@
+	if !(OperationResource{Status: OperationCanceled}).HasTerminated() {
+		t.Errorf("azure: OperationResource#HasTerminated failed to return true for a canceled operation")
+	}
+}
+
+func TestOperationResource_HasTerminatedReturnsTrueIfFailed(t *testing.T) {
+	if !(OperationResource{Status: OperationFailed}).HasTerminated() {
+		t.Errorf("azure: OperationResource#HasTerminated failed to return true for a failed operation")
+	}
+}
+
+func TestOperationResource_HasTerminatedReturnsTrueIfSuccessful(t *testing.T) {
+	if !(OperationResource{Status: OperationSucceeded}).HasTerminated() {
+		t.Errorf("azure: OperationResource#HasTerminated failed to return true for a successful operation")
+	}
+}
+
+func TestOperationResource_HasTerminatedReturnsFalseNotTerminated(t *testing.T) {
+	if (OperationResource{Status: "UnknownStatus"}).HasTerminated() {
+		t.Errorf("azure: OperationResource#HasTerminated returned true for a non-terminal operation")
+	}
+}
+
+func TestOperationResource_GetErrorReturnsErrorIfCanceled(t *testing.T) {
+	if (OperationResource{Status: OperationCanceled}).GetError() == nil {
+		t.Errorf("azure: OperationResource#GetError failed to return an error for canceled operations")
+	}
+}
+
+func TestOperationResource_GetErrorReturnsErrorIfFailed(t *testing.T) {
+	if (OperationResource{Status: OperationFailed}).GetError() == nil {
+		t.Errorf("azure: OperationResource#GetError failed to return an error for failed operations")
+	}
+}
+
+func TestOperationResource_GetErrorDoesReturnUnlessFailedOrCanceled(t *testing.T) {
+	if (OperationResource{Status: OperationSucceeded}).GetError() != nil {
+		t.Errorf("azure: OperationResource#GetError return an error for operation that was not canceled or failed")
+	}
+}
+
+func TestDoPollForAsynchronous_IgnoresUnspecifiedStatusCodes(t *testing.T) {
+	client := mocks.NewSender()
+
+	r, _ := autorest.SendWithSender(client, mocks.NewRequest(),
+		DoPollForAsynchronous(time.Duration(0), time.Duration(0)))
+
+	if client.Attempts() != 1 {
+		t.Errorf("azure: DoPollForAsynchronous polled for unspecified status code")
+	}
+
+	autorest.Respond(r,
+		autorest.ByClosing())
+}
+
+func TestDoPollForAsynchronous_PollsForSpecifiedStatusCodes(t *testing.T) {
+	client := mocks.NewSender()
+	client.AppendResponse(newAsynchronousResponse())
+
+	r, _ := autorest.SendWithSender(client, mocks.NewRequest(),
+		DoPollForAsynchronous(time.Millisecond, time.Millisecond))
+
+	if client.Attempts() != 2 {
+		t.Errorf("azure: DoPollForAsynchronous failed to poll for specified status code")
+	}
+
+	autorest.Respond(r,
+		autorest.ByClosing())
+}
+
+func TestDoPollForAsynchronous_CanBeCanceled(t *testing.T) {
+	cancel := make(chan struct{})
+	delay := 5 * time.Second
+
+	client := mocks.NewSender()
+	client.AppendResponse(newAsynchronousResponse())
+	client.AppendAndRepeatResponse(newOperationResourceResponse("Busy"), -1)
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	start := time.Now()
+	go func() {
+		wg.Done()
```

Comment:
@ahmetalpbalkan I think you may be missing the necessary execution flow.

The code tests that DoPollForAsynchronous will exit if canceled (that is, the http.Request.Cancel channel is closed). This requires that the code execute DoPollForAsynchronous on a "background thread" (as it were) while the "main thread" issues a cancel (that is, closes the channel). It uses a WaitGroup to pause the main thread so the goroutine has a chance to run. The goroutine must release the main thread *before* it enters its own "spin" loop; that is, wg.Done must called *before* SendWithSender.

I hope this clarifies the test.